lib_device_control/host/device_access_usb.c

```diff
@@ -175,6 +176,64 @@ control_read_command(control_resid_t resid, control_cmd_t cmd,
   return CONTROL_SUCCESS;
 }
 
+control_ret_t list_connected_devices(int vendor_id)
+{
+  uint32_t pid_list[MAX_NUM_DEVICES];
+  uint8_t num_devices = 0;
+  uint32_t current_pid = 0;
+  #ifdef _WIN32
+  usb_init();
+  usb_find_busses(); /* find all busses */
+  usb_find_devices(); /* find all connected devices */
+  for (struct usb_bus *bus = usb_get_busses(); bus && !devh; bus = bus->next) {
+    for (struct usb_device *dev = bus->devices; dev; dev = dev->next) {
+      if (dev->descriptor.idVendor == vendor_id) {
+        devh = usb_open(dev);
+        if (!devh) {
+          fprintf(stderr, "Failed to open device with Product ID: %#04x\n", dev->descriptor.idProduct);
+          return CONTROL_ERROR;
+        }
+        current_pid = dev->descriptor.idProduct;
+  #else
+  int ret = libusb_init(NULL);
+  if (ret < 0) {
+    fprintf(stderr, "failed to initialise libusb\n");
+    return CONTROL_ERROR;
+  }
+
+  libusb_device **devs = NULL;
+  int num_dev = libusb_get_device_list(NULL, &devs);
+  struct libusb_device_descriptor desc;
+  for (int dev_idx=0; dev_idx<num_dev; dev_idx++) { 
+    libusb_get_device_descriptor(devs[dev_idx], &desc);
+    if (desc.idVendor == vendor_id) {
+        current_pid = desc.idProduct;
+  #endif        
+        if (num_devices+1>=MAX_NUM_DEVICES) {
+          fprintf(stderr, "Maximum number of connected devices reached: %d\n", MAX_NUM_DEVICES);
+          return CONTROL_ERROR;
+        }
+        pid_list[num_devices] = current_pid;
+        num_devices++;
```

Comment:
Abstracting lines 212 - 217 in a way that combines them with 228 - 232 might work. The resulting function could be used in both _WIN32 and non _WIN32 implementations and would remove the need for a fixed size buffer.